protocol/__init__.py

```diff
@@ -0,0 +1,136 @@
+"""
+Functions for interacting with the computable protocol
+"""
+import os
+import logging.config
+from time import sleep
+from web3 import Web3
+from computable.contracts import Datatrust, Voting
+from computable.helpers.transaction import call, send
+
+import settings
+
+LOGGING_CONFIG = os.path.join(settings.ROOT_DIR, 'logging.conf')
+logging.config.fileConfig(LOGGING_CONFIG)
+log = logging.getLogger()
+
+class Protocol():
+    """
+    Class to manage all interactions with the deployed contracts
+    """
+    def __init__(self):
+        self.datatrust = None
+
+    def init_protocol(self, rpc_path, datatrust_contract, datatrust_host, voting_contract, datatrust_key, datatrust_wallet):
+        """
+        :param rpc_path: URL to the RPC provider for the network
+        :param datatrust_contract: Deployed address of the datatrust contract
+        :param datatrust_host: URL of this application. Used to submit self as datatrust
+        :param voting_contract: Deployed address of voting contract
+        :param datatrust_key: Private key for wallet
+        :param datatrust_wallet: Wallet address
+        """
+        self.w3 = self.initialize_web3(rpc_path)
+        self.datatrust_contract = datatrust_contract
+        self.datatrust_host = datatrust_host
+        self.voting_contract = voting_contract
+        self.datatrust_key = datatrust_key
+        self.datatrust_wallet = datatrust_wallet
+
+    def initialize_web3(self, rpc_path):
+        """
+        Setup the web3 provider
+        """
+        w3 = Web3(Web3.HTTPProvider(rpc_path))
+        log.info('w3 provider set')
+        return w3
+
+    def initialize_datatrust(self):
+        """
+        Confirm or create role as datatrust backend in protocol
+        """
+        log.info('Getting current datatrust address from network')
+        self.datatrust = Datatrust(self.datatrust_wallet)
+        self.datatrust.at(self.w3, self.datatrust_contract)
+
+        backend = call(self.datatrust.get_backend_address())
+        if backend == self.datatrust_wallet:
+            log.info('This server is the datatrust host. Resolving registration')
+            resolve = send(
+                self.w3, 
+                self.datatrust_key, 
+                self.datatrust.resolve_registration(
+                    self.w3.sha3(text=self.datatrust_host), {'from': self.datatrust_wallet}
```

Comment:
you don't need to send `from` if `datatrust_wallet` is the address you created the Datatrust HOC with (which i think _is_ the case)

Reply:
Done